decompiler/pipeline/controlflowanalysis/restructuring_commons/condition_based_refinement.py

```diff
@@ -164,7 +163,11 @@ def _is_subexpression_of_cnf_formula(term: LogicCondition, expression: LogicCond
         """
         if expression.is_true or expression.is_false:
             return False
-        if term.is_equivalent_to(expression):
+        if term.is_equal_to(expression):
+            return True
+        if not expression.does_imply(term):
+            return False
+        elif term.does_imply(expression):
             return True
         if expression.is_negation or expression.is_symbol:
             return False
```

Comment:
Could we do something like this here? I'm afraid of someone adding even more cases to this if-construct
```python
if any(...):
    return False
if any(...):
    return True
```

Reply:
The problem is that the order is important. 
The last if in line 172 only should return false if expression and term are not equivalent.

The `is_equivalent` does the three things of lines 166 to 170. We first check for equal_to. If this is not the case we check both equivalent directions. Now, `expression = term ^ sth` can only be the case if expression implies terms. So if the equivalence fails due to this, we can stop here and return false. Otherwise, we check whether term implies expression. Then, they are equivalent and we return true. So the order is really important

Reply:
well, given this is the case, I'd rather opt for the old solution

Reply:
But the computation of all(any(...)) is very time-consuming and we do all this computation also in the old version, but we interrupt earlier if it is not equivalent due to the one reason.

Reply:
I did a last try to make it better. I really think the runtime improvement is worth having this change.